pensieve/statistics.py

```diff
@@ -0,0 +1,160 @@
+        return self
+
+
+@attr.s(auto_attribs=True)
+class StatisticResultCollection:
+    """
+    Represents a set of statistics result data.
+    """
+
+    data: List[StatisticResult] = []
+
+    def append(self, result: StatisticResult):
+        self.data.append(result)
+
+    def merge(self, result_collection: "StatisticResultCollection"):
+        self.data = self.data + result_collection.data
+
+    def save_to_bigquery(self, client, destination_table, append=True):
+        """Stores the data to a BigQuery table with a defined schema."""
+
+        job_config = bigquery.LoadJobConfig()
+        job_config.schema = [
+            bigquery.SchemaField("metric", "STRING"),
+            bigquery.SchemaField("statistic", "STRING"),
+            bigquery.SchemaField("parameter", "FLOAT64"),
+            bigquery.SchemaField("label", "STRING"),
+            bigquery.SchemaField("ci_width", "FLOAT64"),
+            bigquery.SchemaField("point", "FLOAT64"),
+            bigquery.SchemaField("lower", "FLOAT64"),
+            bigquery.SchemaField("upper", "FLOAT64"),
+        ]
+
+        if append:
+            job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_APPEND
+        else:
+            job_config.write_disposition = bigquery.job.WriteDisposition.WRITE_TRUNCATE
+
+        client.load_table_from_json(self.to_dict()["data"], destination_table)
+
+    def to_dict(self):
+        """Return statistic results as dict."""
+
+        return attr.asdict(self)
+
+
+@attr.s(auto_attribs=True)
+class Statistic:
+    """
+    Abstract representation of a statistic.
+
+    A statistic is a transformation that accepts a table of per-client aggregates and
+    returns a table representing a summary of the aggregates with respect to the branches
+    of the experiment.
+    """
+
+    metrics: List[str]
+    pre_treatments: List[PreTreatment]
+
+    @classmethod
+    def name(cls):
+        return __name__  # todo: snake case names?
+
+    def apply(self, df: DataFrame) -> "StatisticResultCollection":
+        """Run statistic on provided dataframe."""
+
+        data = df
+        for pre_treatment in self.pre_treatments:
+            data = pre_treatment.apply(data)
+
+        col = StatisticResultCollection([])
+
+        for metric in self.metrics:
+            if metric in df:
+                col.merge(self.transformation(df, metric))
+
+        return col
+
+    def transformation(self, df: DataFrame, metric: str) -> "StatisticResultCollection":
+        raise NotImplementedError("Statistic subclasses must override transformation()")
+
+    @classmethod
+    def from_config(cls, config_dict: Dict[str, Any]):  # todo: plug in config file support
+        """Create a class instance with the specified config parameters."""
+        return cls(**config_dict)
+
+
+@attr.s(auto_attribs=True)
+class BootstrapOneBranch(Statistic):
+    num_samples: int = 100
+    summary_quantiles: Tuple[int] = (0.5)
+    confidence_interval: float = 0.95
```

Comment:
The "summary quantiles" that bootstrap_one_branch wants are the critical points for the confidence interval over the sampled statistic. For a 95% interval, that should be (0.025, 0.975).